Animate the site icon element between view and edit in the site editor

mtias opened the pull request:
Polish the animation details around the site icon and title between canvas `view` and `edit`. It makes the icon grow to fill the square container when going into "edit" mode and shrink back when going into "view" mode. It also sets some better fallback handling when the icon is loading.

mtias committed:
Add animation to the site icon element in the header.

Grow to fill the square container when going into "edit"
mode and shrink back when going into "view" mode.

mtias committed:
Balance W icon size and intial animation state.

richtabor commented on packages/edit-site/src/components/site-icon/index.js:
```diff
@@ -41,7 +41,7 @@ function SiteIcon( { className } ) {
 		<Icon
 			className="edit-site-site-icon__icon"
-			size="32px"
+			size="48px"
```
Feels a bit large when its the WordPress icon. 

As-is: 

https://user-images.githubusercontent.com/1813435/223741778-693f0081-8cab-4f44-aa77-bbb5bd294b38.mp4

But if we use `40px`, and `0.65` for scale we have this: 

https://user-images.githubusercontent.com/1813435/223742852-8a3c5a2f-1d74-4af1-b30d-ff194133f5fe.mp4

richtabor replied:
Although (with the suggestion above) the site icon is perhaps a bit big.

https://user-images.githubusercontent.com/1813435/223743492-5e3b2d5b-7e59-40fe-8f1f-0ea7ca72d029.mp4

richtabor commented:
Idea: An alternative to the zoom in on hover, would be to zoom out instead — insinuating the result of the click. 

https://user-images.githubusercontent.com/1813435/223746149-ed3c2a7f-e73b-42bd-acb3-b7493472e910.mp4

jasmussen commented:
This is cool, I like it. Testing with a site logo:

![edit](https://user-images.githubusercontent.com/1204802/223748964-d7438877-4eec-4c71-96d9-fb7c3fe960b3.gif)

The focus style is rectangular, I think that may be fixed by a rebase (see #48918). 

This isn't present in the post editor, maybe that's fine?

<img width="264" alt="Screenshot 2023-03-08 at 16 02 30" src="https://user-images.githubusercontent.com/1204802/223749081-73f0c3db-89f0-4518-97d9-9967c18b2694.png">

There's a detail in the full-size version, here's a nearest neighbor upscaled version to showcase it:

<img width="840" alt="Screenshot 2023-03-08 at 16 03 55" src="https://user-images.githubusercontent.com/1204802/223749319-51de3bd7-94b0-4866-95bf-06ae04a0412c.png">

Note the missing bottom border. The toolbar is 60px tall, the W back button is 61px to overlap the 1px border below. This is so that border can be painted using transparency to better fit in various theme contexts. A simple fix, probably, is to only allow the fullsize site logo to be 60px tall.

mtias commented:
@rich I tried the zoomed out on hover (it was the original one we in #33132!) but after trying it for a few days it felt like it was clashing a bit with the zoom out effect during the transition. Curious how it feels for you after more repeated use.

mtias committed:
Simplify layout of site hub elements.

mtias committed:
Adjust timings so that title fades better according to the toolbar.

mtias commented:
Made some tweaks to orchestrate some of the fades better.

youknowriad committed:
Fix site icon loading

mtias committed:
Address issue with layout animation:

- Resolve problem where title element would shift on initial render.
- Improve orchestration so there's no visible overlap of elements.

mtias committed:
Set up shape fallback while loading site icon.

mtias commented:
**Note:** there's a bug to address where items in the toolbar are not clickable because the site title element lingers on top.

richtabor commented:
The animation that occurs when it shinks + the sidebar comes into view is really nice, but it's still a bit unexpected that the site icon would grow and not be soft-cropped to its fill area: 

https://user-images.githubusercontent.com/1813435/228984171-c415207d-efe9-4885-896c-b0714c20a0e5.mp4

I still keep coming back to the original shrink, then shrink more vibe (what was there a while ago in a sense), as it hints at what's to come (the sidebar), instead of animation for the sake of animation. 

Noting that the animation for the Block Editor site icon look really nice: 

https://user-images.githubusercontent.com/1813435/228985243-2b971f3b-cf64-40e3-b5ee-0fc0d93de4dd.mp4

richtabor commented:
On another note, how strongly do you feel these two areas should be visible different (one with the dark surround, the other full image)? It seems like they are different to help indicate that one is the Site Editor and the other is the Block Editor, though it's not very clear one or the other (other than perhaps a learned effort).

jasmussen commented:
Taking this for a spin:
![move](https://user-images.githubusercontent.com/1204802/229051289-89d9d40e-fd6a-4d7b-9afc-784ad941697b.gif)

I like this a lot. I hear Rich's concern that it scales up further on hover, but it bugs me less, perhaps because my mountain picture is harder to glean details from. I suspect the image you use here is going to make a big difference in how this is perceived too. Not in a problematic way, necessarily, just that different. Here's with no logo, for example:

![no logo](https://user-images.githubusercontent.com/1204802/229052428-7d60d553-62a6-4dbf-972f-eb16db6dcd61.gif)

I'm happy enough to land this. My main feedback would be to crop the 60x60 square so that when you hover over it in edit view, it doesn't grow wider. At the moment it does.

jasmussen commented:
One thing I like about this branch, and why I think it's worth landing, is that switching from this branch to test another, I immedialy realized how much more complex this silhouette is:

<img width="371" alt="Screenshot 2023-03-31 at 09 52 16" src="https://user-images.githubusercontent.com/1204802/229059473-26ee840c-47d2-4359-b247-a6744189dd1f.png">

The rounded icon inside a dark square, next to a busy toolbar. If it's just the full image, it's more of a stamp, rather than a shape.

mtias committed:
Adjust width, height, and overflow to better render the icon.

mtias commented:
Pushed a tweak to improve the cropping. @richtabor I think wherever we land we should reconcile the two editors. I also like the slight zoom-out on the post editor one, though I think it works better because it's on mouseout there, so the slight bounce makes sense.

richtabor commented:
The tweak looks good to me. ✨

apeatling commented:
This is looking great, this kind of polish makes a lot of difference. 👏

pablohoneyhoney commented:
Another nuance is to amplify the image within the editor mode (exceeding the full bleed a bit) so on hover we can do a slight zoom out (without loosing the full bleed) indicating the behavior of zooming out to view mode. 

In a way the reverse of zooming in that the site icon has now on editor view.

mtias committed:
Change order of zoom operations and adjust bottom border.

mtias commented:
I'm trying the inverse zoom effect, though we'd probably have blurriness on the resting state in the current implementation, we would need to update the default bleed to be something like 64px as scale(1) and just go lower in the two steps hover and toggled.

apeatling reviewed (commented):
**Edit:** I realize that much of this is out of scope for this PR, just making a note since it's part of a bigger transition. :D

-------

Tested this more thoroughly. I think it's on its way to being really smooth. There's a couple of things that stood out to me that add to some awkwardness. I tried to emphasize the following points in the video.

1. The canvas frame on resize down and up shifts dramatically to the right, which causes the right edge of the canvas to feel like it's delayed in making it to its final destination. It would look much better if the right at left edges moved directly from their start positions to their ending positions like a true zoom in/out.

2. The top white bar and the site icon immediately start moving up/down and resizing when you click the icon. Because the canvas has so far to move in comparison, it makes it feel more disjointed. I wonder if a slight delay on the icon and top bar to let the canvas lead the change would make it feel more cohesive.

https://user-images.githubusercontent.com/1464705/229859653-691ab05e-6854-481b-a550-cad10a34285b.mp4

richtabor commented:
Here's what I have. The frame is a bit funky, but the icon element looks nice — maybe a tad quick on expanding back to fill the space. 

On a different note, I half-expected that the Site Icon function as a toggle: clicking the icon would toggle between view and edit (instead of clicking on the frame). 🤔

https://user-images.githubusercontent.com/1813435/230140031-9b4b50fe-403a-4cde-846c-afda7db1fa18.mp4

mtias commented:
@apeatling yes, we need to follow up with improvements to the rest of the frame, just trying to keep this one focused on the icon. I also don't think the detached header bar is working quite well in practice, there just seems like too many elements at play, so we should do a pass at that too and simplify.

@richtabor it was a toggle for a brief period in time, but the general idea is now that it would always take you further up in the menu hierarchy, like the iPhone home button used to do; we are just not there yet in the entire admin experience allowing for it to work that way.

mtias committed:
Ensure we are always scaling downwards.

youknowriad commented on packages/edit-site/src/components/site-hub/index.js:
```diff
@@ -41,32 +43,33 @@ const SiteHub = forwardRef( ( props, ref ) => {
+	const controls = useAnimation();
+	useEffect( () => {
+		controls.set( { layout: true } );
+	} );
```
You want to do this, only on mount right? I guess you're missing `[]` as a second argument of useEffect

mtias replied:
yeah, the linter was complaining

mtias committed:
Account for double border in the header.

mtias committed:
Remove unnecessary lines.

SaxonF commented:
> there just seems like too many elements at play, so we should do a pass at that too and simplify.

I agree with this. Are we still totally against using different buttons for editor and admin? This would really help to simplify the transition as you wouldn't have to cross surfaces in a weird way. We could use WordPress logo (or just back arrow) for editor, and site icon for admin. We can generate a default site icon if there isn't one. Identifying what site you're on isn't really a huge issue since the site is in view and gives the best indication. The separate icons might also help their actions feel a bit more predictable vs changing depending on whether you're in editor or not.

Animating header down in frame might also help with the perception of too many moving parts.

The below is using a simple tween animation but might feel better with spring. You can try it out on branch `try/frame-resizer`

https://user-images.githubusercontent.com/1072756/231667822-04d4a854-4e60-4222-b819-766ac4a7ec0d.mp4

youknowriad reviewed (approved):
LGTM in terms of code changes. The animation feels smooth but your call on whether that's the best approach.

ciampo commented:
Gave this a quick test, the W logo is not looking centered for me

https://user-images.githubusercontent.com/1083581/231759829-071da246-9fad-4c4a-a6de-0f9c6de0f950.mp4




While debugging the layout of the W icon, it looks like there are a few nested wrappers with slightly different sizes (some are 58x60 px, some 64x60px, some 64x64px):


https://user-images.githubusercontent.com/1083581/231762016-439f279e-263c-4902-9750-af81b0fa6ca1.mp4

It looks like resizing those wrappers, and adding `align-items: center; justify-content: center` on the flex icon wrapper would improve the layout.

ciampo commented on packages/edit-site/src/components/site-hub/index.js:
```diff
@@ -41,32 +43,33 @@ const SiteHub = forwardRef( ( props, ref ) => {
 
+	const controls = useAnimation();
+	useEffect( () => {
+		controls.set( { layout: true } );
```
Is there a specific reason why can't we set the `layout` prop directly on the motion components ? Maybe there's a way to avoid using the `useAnimation` hook altogether and simplify the code?

mtias replied:
I couldn't find one that worked.

ciampo reviewed (commented):
I was looking for `useAnimation` docs and I couldn't find them. This led me to dig a bit deeper:

- it turns out that `useAnimation` [was renamed to `useAnimationControls` in version `6.4.1`](https://github.com/framer/motion/blob/8b37340b83135b5f03108a77212191efe194ff73/CHANGELOG.md?plain=1#L1024)
- The [`useAnimationControls` docs](https://www.framer.com/motion/use-animation-controls/), they are suggesting to use `useAnimate` instead [(which was added recently in version `10.5.0`)](https://github.com/framer/motion/blob/8b37340b83135b5f03108a77212191efe194ff73/CHANGELOG.md?plain=1#L155), hinting that `useAnimationControls` will soon be deprecated

In case we couldn't avoid using the `useAnimation` hook ([comment](https://github.com/WordPress/gutenberg/pull/48886#discussion_r1165486969)), I wonder if it would be worth looking into updating framer motion first, and then using `useAnimate` instead ?

mtias committed:
Fix centering of W icon.

mtias committed:
Revert useAnimation() as it no longer works in later versions

ciampo committed:
Tweak button sizing & border in edit mode

ciampo commented:
I noticed a white gap above the site logo when in editing mode.

![Screenshot 2023-04-14 at 15 39 02](https://user-images.githubusercontent.com/1083581/232059280-325ae5d7-66ad-41e0-9f31-38873ff636e9.jpg)

I pushed [a commit](https://github.com/WordPress/gutenberg/pull/48886/commits/fe6219021a4210395c660d3c1d6b5ef2af793ac3) that slightly reworks the height of the hub and the grey bottom border in editing mode, which should remove the gap.


@mtias , I guess we can remove the framer-motion dependency update from this PR ?

aristath commented on packages/edit-site/src/components/site-hub/index.js:
```diff
@@ -41,20 +42,15 @@ const SiteHub = forwardRef( ( props, ref ) => {
-		  };
+	const siteIconButtonProps = {
+		label: __( 'Open Admin Sidebar' ),
+		onMouseDown: () => {
```
I'm late to the party since this has already been merged, but is there a reason this was changed from `onClick` to `onMouseDown`? 
As the name implies, `onMouseDown` only works with the mouse (or other pointing devices), and that breaks keyboard navigation 🤔

mtias replied:
That's an oversight of one of the rebases — the intention was to allow triggering the layout switch faster before click release.

t-hamano commented on packages/edit-site/src/components/site-hub/index.js:
```diff
@@ -41,20 +42,15 @@ const SiteHub = forwardRef( ( props, ref ) => {
-	const isBackToDashboardButton = canvasMode === 'view';
-	const showLabels = canvasMode !== 'edit';
-	const siteIconButtonProps = isBackToDashboardButton
-		? {
-				href: dashboardLink || 'index.php',
-				label: __( 'Go back to the Dashboard' ),
-		  }
```
I have discovered that although the site icon is focusable in the browse mode, nothing happens when I click on it. Also, the tooltip labels do not seem to be correct. Was this part intentionally removed?

![view-mode](https://user-images.githubusercontent.com/54422211/236500728-5600db4e-beb9-4bfc-ab8b-e214777b4306.png)

t-hamano replied:
Confirmed that the issue I mentioned was fixed as part of #50296.
